src/java.base/unix/native/libmallochooks/mallochooks.c

```diff
@@ -0,0 +1,523 @@
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#include <sys/types.h>
+#include <stddef.h>
+#include <dlfcn.h>
+#include <errno.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <string.h>
+#include <stdbool.h>
+
+#if defined(__APPLE__)
+#include <malloc/malloc.h>
+#else
+#include <malloc.h>
+#endif
+
+#include "mallochooks.h"
+
+// The log level. 0 is none, 1 is basic logging.
+#define LOG_LEVEL 0
+
+// If > 0 we sync after each write.
+#define SYNC_WRITE 0
+
+void write_safe(int fd, char const* buf, size_t len) {
+  int errno_backup = errno;
+  size_t left = len;
+  ssize_t result;
+
+  while ((result = write(fd, buf, left)) > 0) {
+    buf += result;
+    left -= result;
+  }
+
+#if SYNC_WRITE > 0
+  fsync(fd);
+#endif
+
+  errno = errno_backup;
+}
+
+static void print_error(char const* msg) {
+  write_safe(2, msg, strlen(msg));
+}
+
+
+static void unexpected_call() {
+  print_error("Uninitialized function called. libmallochooks.so must be the first preloaded library.\n");
+  exit(1);
+}
+
+// The tag for malloc functions which should be loaded by dl_sym.
+#define LOAD_DYNAMIC  ((void*) unexpected_call)
+
+#if defined(__APPLE__)
+
+static real_malloc_funcs_t impl = {
+  malloc,
+  calloc,
+  realloc,
+  free,
+  posix_memalign,
+  NULL,
+  NULL,
+  valloc,
+  NULL,
+  (malloc_size_func_t*) malloc_size
+};
+
+#define REPLACE_NAME(x) x##_interpose
+
+#elif defined(__GLIBC__)
+
+void* __libc_malloc(size_t size);
+void* __libc_calloc(size_t elems, size_t size);
+void* __libc_realloc(void* ptr, size_t size);
+void  __libc_free(void* ptr);
+void* __libc_memalign(size_t align, size_t size);
+void* __libc_valloc(size_t size);
+void* __libc_pvalloc(size_t size);
+
+static real_malloc_funcs_t impl = {
+  __libc_malloc,
+  __libc_calloc,
+  __libc_realloc,
+  __libc_free,
+  (posix_memalign_func_t*) LOAD_DYNAMIC,
+  __libc_memalign,
+  (aligned_alloc_func_t*) LOAD_DYNAMIC,
+  __libc_valloc,
+  __libc_pvalloc,
+  malloc_usable_size
+};
+
+#else
+#error "Unexpected platform"
+#endif
+
+#if LOG_LEVEL > 0
+
+static void print(char const* str);
+static void print_ptr(void* ptr);
+static void print_size(size_t size);
+
+#else
+
+#define print_ptr(x)
+#define print_size(x)
+#define print(x)
+
+#endif
+
+#ifndef REPLACE_NAME
+#define REPLACE_NAME(x) x
+#endif
+
+static void assign_function(void** dest, char const* symbol) {
+  if (*dest != LOAD_DYNAMIC) {
+    print("Don't need to load '");
+    print(symbol);
+    print("'\n");
+
+    return;
+  }
+
+  print("Resolving '");
+  print(symbol);
+  print("'\n");
+
+  *dest = dlsym(RTLD_NEXT, symbol);
+
+  if (*dest == NULL) {
+    print_error(symbol);
+    print_error(" not found!\n");
+    /* Don't exit if we don't find the method. It is enough to fail
```

Comment:
Because we support a glibc version on x86_64 which doesn't have an aligned_alloc function. So if you run on such an old version, we don't want to fail since the aligned_alloc function will not be called anyway. But on newer versions we still want to be able to track it.

Reply:
Alright then.